### PR TITLE
Bug: nested plugins cannot be found when nested in a shared config (6.0.0)

### PR DESCRIPTION
**Pull request by nevir**
TL;DR: if a shared config package has plugins that npm or yarn decides to install as descendents of it, those plugins cannot be found by the config factory.

This is a reproduction PR to highlight the issue—I haven't determined a fix yet, and would love pointers if you all have ideas!

**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug report ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))

**Tell us about your environment**

* **ESLint Version:** 6.0.0
* **Node Version:** 10.15.1
* **npm Version:** 6.9.0

**Please show your full configuration:**

[See the fixtures for real files](https://github.com/eslint/eslint/tree/84192fb4a5642cb4f22e854ee5330cc15e67801c/tests/fixtures/config-hierarchy/nested-plugins-parser)

<details>
<summary>Package Layout & Configuration</summary>

Package Tree:
```
<root>
  eslint-config-awesome
    awesome-parser
    eslint-plugin-awesome
```

`<root>/.eslintrc.yml`:
```yaml
extends:
  - awesome
```

`eslint-config-awesome/index.js`:
```js
module.exports = {
  parser: 'awesome-parser',
  plugins: ['awesome'],
  rules: {
    'awesome/is-awesome': [2]
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue.**

`eslint any-file.js`

**What did you expect to happen?**

Eslint would make it to linting the desired file

**What actually happened? Please include the actual, raw output from ESLint.**

> Error: Failed to load plugin 'awesome' declared in '.eslintrc.yml » eslint-config-awesome': Cannot find module 'eslint-plugin-awesome'

### CODE REVIEW

**Comment by nevir** [comment]
Honoring [`importerPath` in `ConfigArrayFactory#_loadPlugin`](https://github.com/eslint/eslint/pull/11884/commits/0c424bda0b1b74d03d37861dce213e9c0596cfbd) fixes this specific issue, but triggers other failures.

Interestingly, honoring `importerPath` is the behavior for [extended configs](https://github.com/eslint/eslint/blob/0c424bda0b1b74d03d37861dce213e9c0596cfbd/lib/cli-engine/config-array-factory.js#L686), and [parsers](https://github.com/eslint/eslint/blob/0c424bda0b1b74d03d37861dce213e9c0596cfbd/lib/cli-engine/config-array-factory.js#L748). Smells like it might actually be the right path to take, but I don't know this well enough yet…

File: tests/lib/cli-engine/cascading-config-array-factory.js [hunk 1]
```diff
@@ -575,6 +575,24 @@ describe("CascadingConfigArrayFactory", () => {
...
+                });
+                const file = getFixturePath("nested-plugins-parser", "foo.js");
+                const expected = {
+                    parser: getFixturePath("nested-plugins-parser", "node_modules", "eslint-config-awesome", "node_modules", "awesome-parser", "index.js"),
+                    plugins: ["awesome"],
+                    rules: {
+                        "awesome/is-awesome": [2]
+                    }
+                };
+                const actual = getConfig(factory, file);
```
Comment (nevir) [inline tests/lib/cli-engine/cascading-config-array-factory.js]: This throws with:

```
     Error: Failed to load plugin 'example' declared in '../broken/plugins/.eslintrc': Cannot find module 'eslint-plugin-example'
Require stack:
- /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/broken/plugins/.eslintrc
      at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
      at Function.resolve (internal/modules/cjs/helpers.js:32:19)
      at Object.resolve (lib/shared/relative-module-resolver.js:44:50)
      at ConfigArrayFactory._loadPlugin (lib/cli-engine/config-array-factory.js:840:39)
      at names.reduce (lib/cli-engine/config-array-factory.js:729:33)
      at Array.reduce (<anonymous>)
      at ConfigArrayFactory._loadPlugins (lib/cli-engine/config-array-factory.js:725:22)
      at ConfigArrayFactory._normalizeObjectConfigDataBody (lib/cli-engine/config-array-factory.js:551:32)
      at _normalizeObjectConfigDataBody.next (<anonymous>)
      at ConfigArrayFactory._normalizeObjectConfigData (lib/cli-engine/config-array-factory.js:490:20)
      at _normalizeObjectConfigData.next (<anonymous>)
      at createConfigArray (lib/cli-engine/config-array-factory.js:306:25)
      at ConfigArrayFactory.loadInDirectory (lib/cli-engine/config-array-factory.js:399:16)
      at CascadingConfigArrayFactory._loadConfigInAncestors (lib/cli-engine/cascading-config-array-factory.js:293:46)
      at CascadingConfigArrayFactory.getConfigArrayForFile (lib/cli-engine/cascading-config-array-factory.js:238:18)
      at getConfig (tests/lib/cli-engine/cascading-config-array-factory.js:144:22)
      at Context.it (tests/lib/cli-engine/cascading-config-array-factory.js:572:32)
```

Comment (nevir) [inline tests/lib/cli-engine/cascading-config-array-factory.js]: This throws with:

```
     Error: Failed to load plugin 'awesome' declared in '.eslintrc.yml » eslint-config-awesome': Cannot find module 'eslint-plugin-awesome'
Require stack:
- /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/__placeholder__.js
      at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
      at Function.resolve (internal/modules/cjs/helpers.js:32:19)
      at Object.resolve (lib/shared/relative-module-resolver.js:44:50)
      at ConfigArrayFactory._loadPlugin (lib/cli-engine/config-array-factory.js:840:39)
      at names.reduce (lib/cli-engine/config-array-factory.js:729:33)
      at Array.reduce (<anonymous>)
      at ConfigArrayFactory._loadPlugins (lib/cli-engine/config-array-factory.js:725:22)
      at ConfigArrayFactory._normalizeObjectConfigDataBody (lib/cli-engine/config-array-factory.js:551:32)
      at _normalizeObjectConfigDataBody.next (<anonymous>)
      at ConfigArrayFactory._normalizeObjectConfigData (lib/cli-engine/config-array-factory.js:490:20)
      at _normalizeObjectConfigData.next (<anonymous>)
      at ConfigArrayFactory._normalizeObjectConfigDataBody (lib/cli-engine/config-array-factory.js:544:25)
      at _normalizeObjectConfigDataBody.next (<anonymous>)
      at ConfigArrayFactory._normalizeObjectConfigData (lib/cli-engine/config-array-factory.js:490:20)
      at _normalizeObjectConfigData.next (<anonymous>)
      at createConfigArray (lib/cli-engine/config-array-factory.js:306:25)
      at ConfigArrayFactory.loadInDirectory (lib/cli-engine/config-array-factory.js:399:16)
      at CascadingConfigArrayFactory._loadConfigInAncestors (lib/cli-engine/cascading-config-array-factory.js:293:46)
      at CascadingConfigArrayFactory.getConfigArrayForFile (lib/cli-engine/cascading-config-array-factory.js:238:18)
      at getConfig (tests/lib/cli-engine/cascading-config-array-factory.js:144:22)
      at Context.it (tests/lib/cli-engine/cascading-config-array-factory.js:590:32)
```

Comment (nevir) [reply tests/lib/cli-engine/cascading-config-array-factory.js]: Debug output leading up to the failure:

```
  eslint:config-array-factory Loading YAML config file: /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/.eslintrc.yml +0ms
  eslint:config-array-factory Config file found: /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/.eslintrc.yml +4ms
  eslint:config-array-factory Loading {extends:"awesome"} relative to /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/.eslintrc.yml +14ms
  eslint:config-array-factory package.json was not found: Cannot find module 'eslint-config-awesome/package.json'
Require stack:
- /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/.eslintrc.yml +1ms
  eslint:config-array-factory Loaded: eslint-config-awesome (/private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/node_modules/eslint-config-awesome/index.js) +0ms
  eslint:config-array-factory Loading JS config file: /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/node_modules/eslint-config-awesome/index.js +1ms
  eslint:config-array-factory Loading parser "awesome-parser" from /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/node_modules/eslint-config-awesome/index.js +1ms
  eslint:config-array-factory package.json was not found: Cannot find module 'awesome-parser/package.json'
Require stack:
- /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/node_modules/eslint-config-awesome/index.js +1ms
  eslint:config-array-factory Loaded: awesome-parser (/private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/node_modules/eslint-config-awesome/node_modules/awesome-parser/index.js) +0ms
  eslint:config-array-factory Loading plugin "awesome" from /private/var/folders/j6/0rhhpj0x0db7yqhhby7x27_r0000gn/T/eslint/fixtures/config-hierarchy/nested-plugins-parser/node_modules/eslint-config-awesome/index.js +1ms
  eslint:config-array-factory Failed to load plugin 'awesome' declared in '.eslintrc.yml » eslint-config-awesome'. +0ms
```

Comment (nevir) [reply tests/lib/cli-engine/cascading-config-array-factory.js]: Note that the ConfigArrayFactory is trying to resolve plugin 'awesome' relative to the root package, not eslint-config-awesome (but the require stack thinks otherwise)

**Comment by mysticatea** [comment]
Hi. Thank you for your contribution.

But this is by design. All plugins must be in `$CWD/node_modules` because rule ID must be identified to a rule. You can see some long discussion: 

- https://github.com/eslint/eslint/issues/3458
- https://github.com/eslint/rfcs/pull/5
- https://github.com/eslint/rfcs/pull/14

**Comment by nevir** [comment]
Ah, doh :( Thanks for the background 👍 That's a tricky set of constraints to navigate…
